File: golang/maximum_depth_of_n-ary_tree.go
package main

import "golang/model"

// Space : O(n)
// Time  : O(n)

func maxNaryNodeDepth(root *model.NaryNode) int {
    if root == nil {
        return 0
    }
    
    ans := 0
    stack := make([]*model.NaryNode, 0)
    stack = append(stack, root)
    
    for len(stack) > 0 {
        ans++
        temp := make([]*model.NaryNode, 0)
        
        for i := 0; i < len(stack); i++ {
            temp = append(temp, stack[i].Children...)
        }
        
        stack = temp
    }
    
    return ans
}

File: python3/maximum_depth_of_n-ary_tree.py
from .model import NaryNode

# Space : O(n)
# Time  : O(n)


class Solution:
    def maxDepth(self, root: NaryNode) -> int:
        if not root:
            return 0

        ans = 0
        stack = [root]

        while stack:
            ans += 1
            temp = []
            for node in stack:
                if node:
                    temp += node.children
            stack = temp

        return ans
